src/d3a/d3a_core/sim_results/file_export_endpoints.py

```diff
@@ -370,50 +371,83 @@ def _accumulate_energy(self, area, child, is_cep: bool):
                     if trade.buyer is child.name:
                         self._total_energy += trade.offer.energy
 
-    def _export_house_pv_self_consumption(self, area):
-        house_pv_device = list()
-        house_load_device = list()
-        trade_by_pv = 0
-        total_energy_bought = 0
-        traded_from_pv = 0
-        if not _is_house_node(area):
-            return
+    def _accumulate_devices(self, area):
         for child in area.children:
-            if isinstance(child.strategy, PVStrategy):
-                house_pv_device.append(child.name)
-            elif _is_load_node(child) or _is_prosumer_node(child):
-                house_load_device.append(child.name)
+            if _is_producer_node(child):
+                self.producer_list.append(child.name)
+            elif _is_load_node(child):
+                self.consumer_list.append(child.name)
+                self.consumer_area_list.append(child.parent)
+                for time, energy in child.strategy.state.desired_energy_Wh.items():
+                    self.total_energy_demanded += energy
+            elif _is_prosumer_node(child):
+                self.ess_list.append(child.name)
+
+            if child.children:
+                self._accumulate_devices(child)
+
+    def _accumulate_self_consumption(self, trade):
+        if trade.seller_origin in self.producer_list and trade.buyer_origin in self.consumer_list:
+            self.total_self_consumption += trade.offer.energy * 1000
+
+    def _accumulate_self_consumption_buffer(self, trade):
+        if trade.seller_origin in self.producer_list and trade.buyer_origin in self.ess_list:
+            self.self_consumption_buffer += trade.offer.energy * 1000
+
+    def _dissipate_self_consumption_buffer(self, trade):
+        if trade.seller_origin in self.ess_list and \
+                (trade.buyer_origin not in self.consumer_list or
+                 trade.buyer_origin not in self.ess_list) and \
+                self.self_consumption_buffer > 0:
+            if (self.self_consumption_buffer - trade.offer.energy * 1000) > 0:
+                self.self_consumption_buffer -= trade.offer.energy * 1000
+            else:
+                self.self_consumption_buffer = 0
```

Comment:
AFAIU, this branch should never happen, right? Or are you just catching the values close to FLOATING_POINT_TOLERANCE ?

Reply:
If the the ESS sold energy to any external load more than what was recorded in `self_consumption_buffer`, in such a case it means there is no energy in ess which was originated from internal producer and has to be set to zero.

Reply:
Ah, yes, makes sense! Thanks!